Automattic/wp-calypso
# Experiment: add smoke tests to P2 frontend editor using Playwright.

#### Changes proposed in this Pull Request

This PR is an **experiment** to add smoke test coverage to the frontend P2 editor, as requested in #56664.

Key changes:
- set up a P2 test blog, visible to external users (e2eflowtesting users are technically treated as external users).
- implement a TOTP client to permit 2FA (required for any user accessing P2s).
- new class `IsolatedBlockEditorComponent` to interact with non-iframed Gutenberg editor on P2s.
- new spec under `test/e2e/specs/specs-p2` to log in, write a post and ensure post is visible.

#### Testing instructions

No TeamCity build configuration is set to run this spec at this moment.

To test:

1. pull the branch locally.
2. navigate to `test/e2e`.
3. run `yarn jest --verbose specs/specs-p2/p2-post.ts
`yarn jest --verbose specs/specs-p2/p2-post.ts`

The output should be something like:

```
 PASS  specs/specs-p2/p2-post.ts (20.509 s)
  Post
    ✓ Log In (6211 ms)
    ✓ View P2 (3403 ms)
    ✓ Add a Paragraph block (874 ms)
    ✓ Enter text (214 ms)
    ✓ Submit post (165 ms)
    ✓ Validate post submission was successful (5925 ms)

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        20.558 s, estimated 50 s
Ran all test suites matching /specs\/specs-p2\/p2-post.ts/i.
```

Remaining todo items:
- [ ] Set up TeamCity build configuration for P2 tests.
- [ ] File issue for refactoring LoginPage.login and LoginPage.enter2FA.
- [ ] File issue for refactoring GutenbergEditorPage and IsolatedBlockEditorComponent.

Related to #56664.
